File: backend/src/main/java/com/sms/service/CourseService.java
package com.sms.service;

import com.sms.dto.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.List;

public interface CourseService {
    Page<CourseResponse> getAllCourses(String search, Long departmentId, Pageable pageable);
    CourseResponse getCourseById(Long id);
    CourseResponse createCourse(CourseRequest request);
    CourseResponse updateCourse(Long id, CourseRequest request);
    void deleteCourse(Long id);
    void enrollStudent(Long courseId, Long studentId);
    void unenrollStudent(Long courseId, Long studentId);
    List<StudentResponse> getEnrolledStudents(Long courseId);
    List<CourseResponse> getCoursesByStudent(Long studentId);
    List<CourseResponse> getCoursesByTeacher(Long teacherId);
}

File: frontend/src/services/courseService.ts
import api from './api';
import { Course, Student } from '../types';

export const courseService = {
  getAll: async (params?: { search?: string; departmentId?: number; page?: number; size?: number }) => {
    const res = await api.get('/api/courses', { params });
    return res.data;
  },
  getById: async (id: number): Promise<Course> => {
    const res = await api.get<Course>(`/api/courses/${id}`);
    return res.data;
  },
  create: async (data: Partial<Course>): Promise<Course> => {
    const res = await api.post<Course>('/api/courses', data);
    return res.data;
  },
  update: async (id: number, data: Partial<Course>): Promise<Course> => {
    const res = await api.put<Course>(`/api/courses/${id}`, data);
    return res.data;
  },
  delete: async (id: number): Promise<void> => {
    await api.delete(`/api/courses/${id}`);
  },
  enrollStudent: async (courseId: number, studentId: number): Promise<void> => {
    await api.post(`/api/courses/${courseId}/enroll/${studentId}`);
  },
  unenrollStudent: async (courseId: number, studentId: number): Promise<void> => {
    await api.delete(`/api/courses/${courseId}/enroll/${studentId}`);
  },
  getEnrolledStudents: async (courseId: number): Promise<Student[]> => {
    const res = await api.get<Student[]>(`/api/courses/${courseId}/students`);
    return res.data;
  },
  getCoursesByStudent: async (studentId: number): Promise<Course[]> => {
    const res = await api.get<Course[]>(`/api/courses/student/${studentId}`);
    return res.data;
  },
  getCoursesByTeacher: async (teacherId: number): Promise<Course[]> => {
    const res = await api.get<Course[]>(`/api/courses/teacher/${teacherId}`);
    return res.data;
  },
};
